File: implementations/go/crypto/hash.go
package crypto

import (
	"crypto/hmac"
	"crypto/sha256"
	"fmt"
	"io"

	"golang.org/x/crypto/hkdf"
	"golang.org/x/crypto/sha3"
)

// Keccak256 returns the 32-byte Keccak-256 digest of the concatenation of the
// given byte slices. This is the original Keccak padding (Ethereum-style), not
// FIPS-202 SHA3-256.
func Keccak256(parts ...[]byte) []byte {
	h := sha3.NewLegacyKeccak256()
	for _, p := range parts {
		h.Write(p)
	}
	return h.Sum(nil)
}

// Sha256 returns the 32-byte SHA-256 digest of the concatenation of the given
// byte slices.
func Sha256(parts ...[]byte) []byte {
	h := sha256.New()
	for _, p := range parts {
		h.Write(p)
	}
	return h.Sum(nil)
}

// HMACSHA256 returns the HMAC-SHA256 of message under key.
func HMACSHA256(key, message []byte) []byte {
	m := hmac.New(sha256.New, key)
	m.Write(message)
	return m.Sum(nil)
}

// HKDFSHA256 derives length bytes of key material from the input keying
// material using HKDF (extract-and-expand) with SHA-256. salt and info may be
// nil. It panics only if length exceeds HKDF's 255*HashLen ceiling, which is a
// programming error for the fixed-size derivations used by the protocol.
func HKDFSHA256(ikm, salt, info []byte, length int) []byte {
	r := hkdf.New(sha256.New, ikm, salt, info)
	out := make([]byte, length)
	if _, err := io.ReadFull(r, out); err != nil {
		panic(fmt.Sprintf("crypto: HKDFSHA256 length %d too large: %v", length, err))
	}
	return out
}

File: js/packages/sdk/src/crypto/hash.ts
import { keccak_256 } from "@noble/hashes/sha3.js";
import { sha256 as _sha256, sha512 as _sha512 } from "@noble/hashes/sha2.js";
import { hmac } from "@noble/hashes/hmac.js";
import { hkdf } from "@noble/hashes/hkdf.js";
import { concatBytes } from "@noble/hashes/utils.js";

/**
 * Keccak-256 (original Keccak padding, as used by Ethereum) of the concatenated
 * parts. Distinct from FIPS-202 SHA3-256.
 */
export function keccak256(...parts: Uint8Array[]): Uint8Array {
  return keccak_256(concatBytes(...parts));
}

/** SHA-256 of the concatenated parts. */
export function sha256(...parts: Uint8Array[]): Uint8Array {
  return _sha256(concatBytes(...parts));
}

/** SHA-512 of the concatenated parts. */
export function sha512(...parts: Uint8Array[]): Uint8Array {
  return _sha512(concatBytes(...parts));
}

/** HMAC-SHA256 of message under key. */
export function hmacSha256(key: Uint8Array, message: Uint8Array): Uint8Array {
  return hmac(_sha256, key, message);
}

/**
 * HKDF (extract-and-expand) with SHA-256, producing `length` bytes. `salt` and
 * `info` are passed verbatim; an empty `salt` is used as an empty HMAC key
 * (matching the Go reference), not zero-filled.
 */
export function hkdfSha256(
  ikm: Uint8Array,
  salt: Uint8Array,
  info: Uint8Array,
  length: number,
): Uint8Array {
  return hkdf(_sha256, ikm, salt, info, length);
}
